test/Microsoft.AspNet.Razor.Runtime.Test/Runtime/TagHelpers/TagHelperExecutionContextTest.cs

```diff
@@ -65,30 +80,73 @@ public void ParentItems_SetsItemsProperty()
+
+            // Assert
+            Assert.Equal(expectedContent, actualContent.GetContent(new HtmlTestEncoder()));
+        }
 
+        [Theory]
+        [MemberData(nameof(HtmlEncoderData))]
+        public async Task GetChildContentAsync_StartsWritingScopeWithGivenEncoder(HtmlEncoder encoder)
+        {
+            // Arrange
+            HtmlEncoder passedEncoder = null;
+            var executionContext = new TagHelperExecutionContext(
+                "p",
+                tagMode: TagMode.StartTagAndEndTag,
+                items: new Dictionary<object, object>(),
+                uniqueId: string.Empty,
+                executeChildContentAsync: () => Task.FromResult(result: true),
+                startTagHelperWritingScope: encoderArgument => passedEncoder = encoderArgument,
+                endTagHelperWritingScope: () => new DefaultTagHelperContent());
+
+            // Act
+            await executionContext.GetChildContentAsync(useCachedResult: true, encoder: encoder);
+
+            // Assert
+            Assert.Same(encoder, passedEncoder);
+        }
+
+        [Theory]
+        [MemberData(nameof(HtmlEncoderData))]
+        public async Task GetChildContentAsync_CachesValue(HtmlEncoder encoder)
+        {
+            // Arrange
+            var executionCount = 0;
+            var executionContext = new TagHelperExecutionContext(
+                "p",
+                tagMode: TagMode.StartTagAndEndTag,
+                items: new Dictionary<object, object>(),
+                uniqueId: string.Empty,
+                executeChildContentAsync: () =>
+                {
+                    executionCount++;
                     return Task.FromResult(result: true);
                 },
-                startTagHelperWritingScope: () => { },
-                endTagHelperWritingScope: () => defaultTagHelperContent);
+                startTagHelperWritingScope: _ => { },
+                endTagHelperWritingScope: () => new DefaultTagHelperContent());
 
             // Act
-            var content1 = await executionContext.GetChildContentAsync(useCachedResult: true);
-            var content2 = await executionContext.GetChildContentAsync(useCachedResult: true);
+            var content1 = await executionContext.GetChildContentAsync(useCachedResult: true, encoder: encoder);
+            var content2 = await executionContext.GetChildContentAsync(useCachedResult: true, encoder: encoder);
 
             // Assert
-            Assert.Equal(expectedContent, content1.GetContent(new HtmlTestEncoder()));
-            Assert.Equal(expectedContent, content2.GetContent(new HtmlTestEncoder()));
+            Assert.Equal(1, executionCount);
         }
 
-        [Fact]
-        public async Task GetChildContentAsync_CanExecuteChildrenMoreThanOnce()
+        [Theory]
+        [MemberData(nameof(HtmlEncoderData))]
+        public async Task GetChildContentAsync_CachesValuePerEncoder(HtmlEncoder encoder)
```

Comment:
Should add one more variation of this where you call `GetChildContentAsync` twice for each encoder and ensure that `executionCount` remains 2.